File: frontend/src/services/menuService.js
import api from './api';
import { dataStorage } from '../utils/dataStorage';

export const menuService = {
  getTodayMenu: async () => {
    try {
      const res = await api.get('/menus/today');
      return res;
    } catch {
      const all = dataStorage.getMenu();
      const days = ['SUNDAY', 'MONDAY', 'TUESDAY', 'WEDNESDAY', 'THURSDAY', 'FRIDAY', 'SATURDAY'];
      const today = days[new Date().getDay()];
      return { data: all[today] || { breakfast: [], lunch: [], dinner: [] } };
    }
  },

  getWeeklyMenu: async () => {
    try {
      const res = await api.get('/menus/weekly');
      return res;
    } catch {
      return { data: dataStorage.getMenu() };
    }
  },

  getMenuByDay: async (day) => {
    try {
      const res = await api.get(`/menus/${day}`);
      return res;
    } catch {
      const all = dataStorage.getMenu();
      return { data: all[day.toUpperCase()] || { breakfast: [], lunch: [], dinner: [] } };
    }
  },

  updateMenu: async (day, data) => {
    try {
      const res = await api.put(`/menus/${day}`, data);
      dataStorage.updateDayMenu(day, data);
      return res;
    } catch {
      const updated = dataStorage.updateDayMenu(day, data);
      return { data: updated[day.toUpperCase()] };
    }
  },

  updateWeeklyMenu: async (data) => {
    try {
      const res = await api.put('/menus/weekly', data);
      Object.keys(data).forEach((d) => dataStorage.updateDayMenu(d, data[d]));
      return res;
    } catch {
      Object.keys(data).forEach((d) => dataStorage.updateDayMenu(d, data[d]));
      return { data: dataStorage.getMenu() };
    }
  },
};

File: backend/src/main/java/com/northmess/service/MenuService.java
package com.northmess.service;

import com.northmess.entity.Menu;
import com.northmess.entity.enums.DayOfWeekEnum;
import com.northmess.repository.MenuRepository;
import com.northmess.security.UserPrincipal;
import java.time.DayOfWeek;
import java.time.LocalDate;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class MenuService {

    private final MenuRepository menuRepository;

    public Menu getTodayMenu() {
        DayOfWeekEnum today = mapDay(LocalDate.now().getDayOfWeek());
        return menuRepository.findByDay(today)
            .orElseGet(() -> emptyMenu(today));
    }

    public List<Menu> getWeeklyMenu() {
        List<Menu> menus = new ArrayList<>();
        for (DayOfWeekEnum day : DayOfWeekEnum.values()) {
            menus.add(menuRepository.findByDay(day).orElseGet(() -> emptyMenu(day)));
        }
        menus.sort(Comparator.comparingInt(menu -> menu.getDay().ordinal()));
        return menus;
    }

    public Menu getByDay(String day) {
        DayOfWeekEnum dayEnum = DayOfWeekEnum.from(day);
        return menuRepository.findByDay(dayEnum).orElseGet(() -> emptyMenu(dayEnum));
    }

    public Menu updateMenu(String day, Menu request, UserPrincipal principal) {
        DayOfWeekEnum dayEnum = DayOfWeekEnum.from(day);
        Menu menu = menuRepository.findByDay(dayEnum).orElseGet(() -> emptyMenu(dayEnum));
        menu.setDay(dayEnum);
        if (request.getBreakfast() != null) menu.setBreakfast(request.getBreakfast());
        if (request.getLunch() != null) menu.setLunch(request.getLunch());
        if (request.getDinner() != null) menu.setDinner(request.getDinner());
        menu.setUpdatedAt(Instant.now());
        menu.setUpdatedBy(principal.getName());
        return menuRepository.save(menu);
    }

    public List<Menu> updateWeeklyMenu(List<Menu> menus, UserPrincipal principal) {
        List<Menu> saved = new ArrayList<>();
        for (Menu menu : menus) {
            if (menu.getDay() == null) {
                continue;
            }
            saved.add(updateMenu(menu.getDay().name(), menu, principal));
        }
        return saved;
    }

    private Menu emptyMenu(DayOfWeekEnum day) {
        Menu menu = new Menu();
        menu.setDay(day);
        menu.setBreakfast(List.of());
        menu.setLunch(List.of());
        menu.setDinner(List.of());
        return menu;
    }

    private DayOfWeekEnum mapDay(DayOfWeek day) {
        return switch (day) {
            case MONDAY -> DayOfWeekEnum.MONDAY;
            case TUESDAY -> DayOfWeekEnum.TUESDAY;
            case WEDNESDAY -> DayOfWeekEnum.WEDNESDAY;
            case THURSDAY -> DayOfWeekEnum.THURSDAY;
            case FRIDAY -> DayOfWeekEnum.FRIDAY;
            case SATURDAY -> DayOfWeekEnum.SATURDAY;
            case SUNDAY -> DayOfWeekEnum.SUNDAY;
        };
    }
}